From 7978fe5e608467f9d539f300845d223dcfc3e688 Mon Sep 17 00:00:00 2001
From: poram srinivasa rao <poramx.srinivasa.rao@intel.com>
Date: Wed, 15 Oct 2025 15:40:19 +0530
Subject: [PATCH] Fix for Firmware update with Software Manager

This commit implements the firmware update using redfish postman
tool with new service xyz.openbmc_project.Software.Manager which
changed due to upstream changes from the older service
xyz.openbmc_project.Software.BMC.Updater with
redfish-updateservice-use-dbus enabled.

So, this is like firmware update using dbus call and which copies
firmware image to /tmp/images but uses new service
xyz.openbmc_project.Software.Manager instead of
xyz.openbmc_project.Software.BMC.Updater.

Signed-off-by: P Dheeraj Srujan Kumar <p.dheeraj.srujan.kumar@intel.com>
Signed-off-by: poram srinivasa rao <poramx.srinivasa.rao@intel.com>

Upstream-Status: Pending
---
 meson.build        |   6 +-
 update_manager.cpp | 510 ++++++++++++++++++++++++++++++++++++++++-
 update_manager.hpp |  44 ++++
 3 files changed, 556 insertions(+), 4 deletions(-)

diff --git a/meson.build b/meson.build
index ec1f630..1a9484c 100644
--- a/meson.build
+++ b/meson.build
@@ -5,7 +5,7 @@ conf.set_quoted(
     'BMC_INVENTORY_INTERFACE',
     'xyz.openbmc_project.Inventory.Item.Bmc',
 )
-conf.set_quoted('BUSNAME_UPDATER', 'xyz.openbmc_project.Software.BMC.Updater')
+conf.set_quoted('BUSNAME_UPDATER', 'xyz.openbmc_project.Software.Manager')
 conf.set_quoted('DOWNLOAD_BUSNAME', 'xyz.openbmc_project.Software.Download')
 conf.set_quoted('FILEPATH_IFACE', 'xyz.openbmc_project.Common.FilePath')
 conf.set_quoted('INVENTORY_PATH', '/xyz/openbmc_project/inventory/')
@@ -105,6 +105,7 @@ unit_files = [
     'usr-local.mount.in',
     'xyz.openbmc_project.Software.Download.service.in',
     'xyz.openbmc_project.Software.Sync.service.in',
+    'xyz.openbmc_project.Software.Version.service.in',
 ]
 
 image_updater_sources = files('activation.cpp', 'images.cpp', 'item_updater.cpp', 'msl_verify.cpp', 'serialize.cpp', 'update_manager.cpp'
@@ -205,8 +206,7 @@ if get_option('software-update-dbus-interface').allowed()
     unit_files += ['xyz.openbmc_project.Software.Manager.service.in']
 else
     unit_files += [
-        'xyz.openbmc_project.Software.BMC.Updater.service.in',
-        'xyz.openbmc_project.Software.Version.service.in',
+        'xyz.openbmc_project.Software.BMC.Updater.service.in'
     ]
 endif
 
diff --git a/update_manager.cpp b/update_manager.cpp
index 922f2bc..c0f8e99 100644
--- a/update_manager.cpp
+++ b/update_manager.cpp
@@ -1,9 +1,13 @@
 #include "update_manager.hpp"
 
+#include "dbus_helpers.hpp"
 #include "item_updater.hpp"
+#include "pldm.hpp"
+#include "smm_runtime_update.hpp"
 #include "software_utils.hpp"
 #include "version.hpp"
 
+#include <boost/process/v1/child.hpp>
 #include <phosphor-logging/elog-errors.hpp>
 #include <phosphor-logging/elog.hpp>
 #include <phosphor-logging/lg2.hpp>
@@ -30,6 +34,9 @@ using ManifestFail = SoftwareLogging::image::ManifestFileFailure;
 using UnTarFail = SoftwareLogging::image::UnTarFailure;
 using InternalFail = SoftwareLogging::image::InternalFailure;
 using ImageFail = SoftwareLogging::image::ImageFailure;
+std::map<std::string, std::unique_ptr<Version>> versions;
+static constexpr const uint32_t fwupdateImageIdentifier = 0xB6EAFD19;
+static constexpr const uint32_t pldmMagic = 0xF018878C;
 
 void Manager::processImageFailed(sdbusplus::message::unix_fd image,
                                  std::string& id)
@@ -208,6 +215,507 @@ auto Manager::processImage(sdbusplus::message::unix_fd image,
     co_return;
 }
 
+int Manager::verifyImage(const std::filesystem::path imgPath,
+                         std::string& version, std::string& purposeString)
+{
+    uint8_t imgType{};
+    uint32_t imgMagic{};
+    std::array<uint8_t, 2> verData{};
+    constexpr size_t readBufferSize = std::max<size_t>(
+        {sizeof(pfrImgBlock0), sizeof(pldm::PldmPkgHeader),
+         sizeof(smru::CapsuleHeader)});
+
+    std::array<char, readBufferSize> readBuffer = {};
+
+    std::string verPurpose{
+        "xyz.openbmc_project.Software.Version.VersionPurpose."};
+
+    if (std::filesystem::exists(imgPath))
+    {
+        try
+        {
+            std::ifstream imgFile(imgPath, std::ios::binary | std::ios::in);
+
+            if (!imgFile.good())
+            {
+                phosphor::logging::log<phosphor::logging::level::ERR>(
+                    "Image file read failed");
+                return -1;
+            }
+
+            imgFile.read(readBuffer.data(), readBufferSize);
+
+            auto* block0Data =
+                reinterpret_cast<pfrImgBlock0*>(readBuffer.data());
+
+            imgMagic = block0Data->tag;
+
+            if (imgMagic == smru::smruMagic)
+            {
+                // Verify SMM Capusle recieved.
+                // This function fills the Version number if the Capsule is
+                // valid
+                if (!smru::verifyCapsule(ctx.get_bus(), imgPath, version))
+                {
+                    return -1;
+                }
+
+                phosphor::logging::log<phosphor::logging::level::INFO>(
+                    "SMM Capsule Verified");
+
+                purposeString = verPurpose + "Host";
+                return 0;
+            }
+
+            if (htobe32(imgMagic) == pldmMagic)
+            {
+                if (!version.empty())
+                {
+                    version.clear();
+                }
+
+                imgFile.seekg(0, std::ios_base::end);
+
+                const size_t length = imgFile.tellg();
+                constexpr size_t readBytes = 36;
+
+                if (length < readBytes)
+                {
+                    phosphor::logging::log<phosphor::logging::level::ERR>(
+                        "Insufficient file length to read the required "
+                        "bytes");
+                    return -1;
+                }
+
+                imgFile.seekg(0, std::ios::beg);
+
+                std::array<char, readBytes> buffer = {};
+
+                imgFile.read(
+                    buffer.data(),
+                    buffer.size()); // read 36 bytes of PLDM Package Header
+
+                if (!imgFile.good())
+                {
+                    phosphor::logging::log<phosphor::logging::level::ERR>(
+                        "Image file read is not successful");
+                    return -1;
+                }
+
+                if (!std::equal(buffer.begin(),
+                                buffer.begin() + pldm::headerIdLen,
+                                pldm::pldmPkgHeaderId
+                                    .begin())) // comparing 16 bytes of
+                                               // PackageHeaderIdentifier field
+                {
+                    std::string redfishMsgID =
+                        "OpenBMC.0.1.FirmwareUpdateFailed";
+                    sd_journal_send(
+                        "MESSAGE=%s", "Firmware image verification failed",
+                        "PRIORITY=%i", LOG_ERR, "REDFISH_MESSAGE_ID=%s",
+                        redfishMsgID.c_str(), "REDFISH_MESSAGE_ARGS=%s",
+                        "PLDM Image package header identifier check fail",
+                        NULL);
+
+                    return -1;
+                }
+
+                phosphor::logging::log<phosphor::logging::level::INFO>(
+                    "Package header identifier matched");
+                purposeString =
+                    "xyz.openbmc_project.Software.Version.VersionPurpose.PLDM";
+
+                const uint8_t pkgVerStrLen = static_cast<uint8_t>(
+                    buffer[35]); // PackageVersionStringLen byte
+
+                imgFile.seekg(readBytes,
+                              std::ios::beg); // point to the begin of
+                                              // PackageVersionString field
+                                              // i.e. 36th pos
+
+                std::array<char, 255> ver = {};
+                imgFile.read(ver.data(),
+                             pkgVerStrLen); // read PackageVersionString bytes
+
+                if (!imgFile.good())
+                {
+                    phosphor::logging::log<phosphor::logging::level::ERR>(
+                        "Image file read is not successful");
+                    return -1;
+                }
+
+                version.assign(ver.data(), pkgVerStrLen);
+                phosphor::logging::log<phosphor::logging::level::INFO>(
+                    "Package version string value",
+                    phosphor::logging::entry("IMAGE_VERSION=%s",
+                                             version.c_str()));
+
+                imgFile.close();
+                return 0;
+            }
+
+            if (imgMagic != fwupdateImageIdentifier)
+            {
+                phosphor::logging::log<phosphor::logging::level::ERR>(
+                    "Image magic number match failed",
+                    phosphor::logging::entry("IMAGEMAGIC=0x%x", imgMagic));
+
+                return -1;
+            }
+
+            imgType = block0Data->pcType[0];
+
+            phosphor::logging::log<phosphor::logging::level::INFO>(
+                "Image Type", phosphor::logging::entry(
+                                  "IMAGETYPE=0x%x", static_cast<int>(imgType)));
+
+            if (imgType == pfrBMCUpdateCap || imgType == pfrBMCPFM ||
+                imgType == bmcSecureBootRotCap ||
+                imgType == bmcSecureBootOTPCap)
+            {
+                purposeString = verPurpose + "BMC";
+            }
+            else if (imgType == pfrPCHUpdateCap || imgType == pfrPCHPFM ||
+                     imgType == seamlessPCHUpdCap)
+            {
+                purposeString = verPurpose + "Host";
+            }
+            else if (imgType == pfrCPLDUpdateCap ||
+                     imgType == pfrAFMUpdateCap ||
+                     imgType == pfrCmpstCPLDUpdCap ||
+                     imgType == pfrAddToUpdateAFM ||
+                     imgType == pfrPerDeviceAFM ||
+                     imgType == pfrCmpstRetimerUpdCap)
+            {
+                purposeString = verPurpose + "Other";
+            }
+            else
+            {
+                purposeString = verPurpose + "Unknown";
+
+                phosphor::logging::log<phosphor::logging::level::ERR>(
+                    "Unknown image type");
+                return -1;
+            }
+
+            // Verify the complete image
+            std::string mtdUtilfile = "/usr/bin/mtd-util";
+            std::vector<std::string> mtdUtilCmd = {"p", "a"};
+            mtdUtilCmd.push_back(imgPath.string());
+
+            boost::process::v1::child execProg(mtdUtilfile, mtdUtilCmd);
+            execProg.wait();
+            if (execProg.exit_code())
+            {
+                phosphor::logging::log<phosphor::logging::level::ERR>(
+                    "Image authentication failed");
+
+                constexpr const char* redfishMsgID =
+                    "OpenBMC.0.1.GeneralFirmwareSecurityViolation";
+                sd_journal_send("MESSAGE=%s",
+                                "Firmware image authentication failed",
+                                "PRIORITY=%i", LOG_ERR, "REDFISH_MESSAGE_ID=%s",
+                                redfishMsgID, "REDFISH_MESSAGE_ARGS=%s",
+                                "Image authentication check fail", NULL);
+                return -1;
+            }
+
+            phosphor::logging::log<phosphor::logging::level::INFO>(
+                "PFR image",
+                phosphor::logging::entry("PCType=%d", block0Data->pcType[0]),
+                phosphor::logging::entry("VERSION=%s", version.c_str()));
+
+            version = std::to_string(verData[0]) + "." +
+                      std::to_string(verData[1]);
+        }
+        catch (std::exception& e)
+        {
+            phosphor::logging::log<phosphor::logging::level::ERR>(e.what());
+            return -1;
+        }
+    }
+
+    return 0;
+}
+
+int Manager::verifyBmcFullFlashImage(const std::filesystem::path imgPath,
+                                     std::string& version,
+                                     std::string& purposeString)
+{
+    static constexpr const uint32_t pfmOldLayoutOffset = 0x8000;
+    static constexpr const uint32_t pfmNewLayoutOffset = 0x100000;
+    static constexpr const uint32_t rcOldLayoutOffset = 0x02A00000;
+    static constexpr const uint32_t rcNewLayoutOffset = 0x02F00000;
+
+    // Currently, the BMC full image can be structured in 2 ways, OLD-FLASH
+    // layout or NEW-FLASH layout. In order to verify the image (to make sure
+    // the image is intact), PFM and recovery region can be checked to ensure
+    // the PFM magic and PC type is present at the correct offset (Old/New)
+    // (Block 0 check) Ensure the BMC magic and PC type is present at the
+    // correct offset (Old/New) (Block 0 check)
+
+    // Open the full image
+    std::ifstream imgFile(imgPath, std::ios::binary | std::ios::in);
+
+    if (!imgFile.good())
+    {
+        phosphor::logging::log<phosphor::logging::level::ERR>(
+            "Image file read failed");
+        return -1;
+    }
+
+    std::array<char, sizeof(pfrImgBlock0)> buffer = {};
+
+    // Default bmc RC region offset to new
+    uint32_t bmcBlk0Offset = rcNewLayoutOffset;
+
+    // Read PFM at new offset
+    imgFile.seekg(pfmNewLayoutOffset, std::ios::beg);
+    imgFile.read(buffer.data(), buffer.size());
+    auto* pfmBlock0Data = reinterpret_cast<pfrImgBlock0*>(buffer.data());
+
+    // Check for BLOCK 0 magic of PFM signed region
+    if (pfmBlock0Data->tag != fwupdateImageIdentifier)
+    {
+        // If the magic number is not present at new offset, move to old offset
+        bmcBlk0Offset = rcOldLayoutOffset;
+
+        // Read PFM at old offset
+        imgFile.seekg(pfmOldLayoutOffset, std::ios::beg);
+        imgFile.read(buffer.data(), buffer.size());
+        pfmBlock0Data = reinterpret_cast<pfrImgBlock0*>(buffer.data());
+    }
+
+    // If the magic number is not present at old offset as well, image is not
+    // valid
+    if (pfmBlock0Data->tag != fwupdateImageIdentifier)
+    {
+        phosphor::logging::log<phosphor::logging::level::ERR>(
+            "BMC full image - PFM Block 0 TAG verification failed");
+        return -1;
+    }
+
+    // Verify PC type -> must be PFM(0x03)
+    if (pfmBlock0Data->pcType[0] != pfrImgPCType::pfrBMCPFM)
+    {
+        phosphor::logging::log<phosphor::logging::level::ERR>(
+            "BMC full image - Invalid PFM Block 0 PC Type");
+        return -1;
+    }
+
+    // Read the BMC RC region (Old/New determined above)
+    imgFile.seekg(bmcBlk0Offset, std::ios::beg);
+    imgFile.read(buffer.data(), buffer.size());
+    auto* bmcBlock0Data = reinterpret_cast<pfrImgBlock0*>(buffer.data());
+
+    // Verify BLOCK 0 magic for RC region
+    if (bmcBlock0Data->tag != fwupdateImageIdentifier)
+    {
+        phosphor::logging::log<phosphor::logging::level::ERR>(
+            "BMC full image - RC Block 0 TAG verification failed");
+        return -1;
+    }
+
+    // Verify PC type -> must be BMC(0x04)
+    if (bmcBlock0Data->pcType[0] != pfrImgPCType::pfrBMCUpdateCap)
+    {
+        phosphor::logging::log<phosphor::logging::level::ERR>(
+            "BMC full image - Invalid BMC Block 0 PC Type");
+        return -1;
+    }
+
+    // Image is successfully verified.
+    version = "NA";
+    purposeString = "xyz.openbmc_project.Software.Version.VersionPurpose.BMC";
+
+    return 0;
+}
+
+static bool writeToImage(int imageFd, const std::string& destFilePath)
+{
+    int dest_fd =
+        open(destFilePath.c_str(), O_WRONLY | O_CREAT | O_TRUNC, 0644);
+    if (dest_fd == -1)
+    {
+        return false;
+    }
+
+    char buffer[4096];
+    ssize_t bytes_read;
+
+    while ((bytes_read = read(imageFd, buffer, sizeof(buffer))) > 0)
+    {
+        if (write(dest_fd, buffer, bytes_read) != bytes_read)
+        {
+            return false;
+        }
+    }
+
+    if (bytes_read == -1)
+    {
+        return false;
+    }
+
+    return true;
+}
+
+auto Manager::processPFRImage(sdbusplus::message::unix_fd image,
+                              ApplyTimeIntf::RequestedApplyTimes applyTime,
+                              std::string id, std::string objPath)
+    -> sdbusplus::async::task<>
+{
+    debug("Processing PFR image {FD}", "FD", image.fd);
+    fs::path tmpDirPath(std::string{IMG_UPLOAD_DIR});
+    tmpDirPath /= "imageXXXXXX";
+    auto tmpDir = tmpDirPath.string();
+
+    if (!mkdtemp(tmpDir.data()))
+    {
+        error("Error ({ERRNO}) occurred during mkdtemp", "ERRNO", errno);
+        processImageFailed(image, id);
+        report<SoftwareErrors::InternalFailure>(InternalFail::FAIL("mkdtemp"));
+        co_return;
+    }
+
+    std::error_code ec;
+    tmpDirPath = tmpDir;
+    softwareUtils::RemovablePath tmpDirToRemove(tmpDirPath);
+
+    std::string imgFilePath = tmpDirPath.string() + "/fw_download.bin";
+
+    if (!writeToImage(image, imgFilePath))
+    {
+        error("Error occurred during image writing");
+        processImageFailed(image, id);
+        co_return;
+    }
+
+    std::filesystem::path imgPath(imgFilePath);
+
+    if (!std::filesystem::exists(imgPath))
+    {
+        error("File path doesnot exists");
+        co_return;
+    }
+
+    if (isFwupdScriptRunning(ctx.get_bus()))
+    {
+        error("FwupdScript is Running already");
+        co_return;
+    }
+
+    std::string extVer;
+    std::string ver;
+    std::string purposeString;
+
+    std::uintmax_t fSize = std::filesystem::file_size(imgFilePath);
+
+    if (fSize == 64 * 1024 * 1024)
+    {
+        // TODO: Find out how to read version number and also set correct
+        // "purpose" string. Filling with dummy data for now.
+
+        ver = "1";
+        purposeString =
+            "xyz.openbmc_project.Software.Version.VersionPurpose.Host";
+    }
+
+    else if (fSize == 128 * 1024 * 1024)
+    {
+        if (0 != verifyBmcFullFlashImage(imgFilePath, ver, purposeString))
+        {
+            phosphor::logging::log<phosphor::logging::level::ERR>(
+                "Error verifying uploaded image");
+            std::filesystem::remove_all(imgFilePath);
+            co_return;
+        }
+    }
+
+    else if (0 != verifyImage(imgFilePath, ver, purposeString))
+    {
+        phosphor::logging::log<phosphor::logging::level::ERR>(
+            "Error verifying uploaded image");
+        std::filesystem::remove_all(imgFilePath);
+        co_return;
+    }
+
+    if (ver.empty())
+    {
+        phosphor::logging::log<phosphor::logging::level::ERR>(
+            "Empty version from image file");
+        co_return;
+    }
+
+    sdbusplus::xyz::openbmc_project::Software::server::Version::VersionPurpose
+        purpose = Version::VersionPurpose::Unknown;
+
+    try
+    {
+        purpose = Version::convertVersionPurposeFromString(purposeString);
+    }
+    catch (const sdbusplus::exception::InvalidEnumString& e)
+    {
+        phosphor::logging::log<phosphor::logging::level::ERR>(
+            "Error: Failed to convert purpose to enum."
+            " Setting to Unknown.");
+    }
+
+    if (versions.find(id) != versions.end())
+    {
+        phosphor::logging::log<phosphor::logging::level::INFO>(
+            "Software Object with the same version already exists, exiting "
+            "the update",
+            phosphor::logging::entry("VERSION_ID=%s", id.c_str()));
+        co_return;
+    }
+
+    std::filesystem::path imageDirPath(IMG_UPLOAD_DIR);
+    imageDirPath /= id;
+    auto filePath = imageDirPath.string();
+    try
+    {
+        std::filesystem::create_directory(imageDirPath);
+    }
+    catch (const std::filesystem::filesystem_error& fsErr)
+    {
+        phosphor::logging::log<phosphor::logging::level::ERR>(fsErr.what());
+    }
+
+    std::filesystem::path newFileName = imageDirPath / "image-runtime";
+    try
+    {
+        std::filesystem::rename(imgFilePath, newFileName);
+    }
+    catch (const std::filesystem::filesystem_error& fsErr)
+    {
+        phosphor::logging::log<phosphor::logging::level::ERR>(fsErr.what());
+    }
+
+    std::vector<std::string> compatibleNames;
+    auto state = itemUpdater.verifyAndCreateObjects(
+        id, objPath, ver, purpose, extVer, filePath, compatibleNames);
+
+    if (state != ActivationIntf::Activations::Ready)
+    {
+        error("Software image is invalid");
+        processImageFailed(image, id);
+        report<SoftwareErrors::ImageFailure>(
+            ImageFail::FAIL("Image is invalid"),
+            ImageFail::PATH(filePath.c_str()));
+        co_return;
+    }
+
+    if (applyTime == ApplyTimeIntf::RequestedApplyTimes::Immediate ||
+        applyTime == ApplyTimeIntf::RequestedApplyTimes::OnReset)
+    {
+        itemUpdater.requestActivation(id);
+    }
+
+    updateInProgress = false;
+    close(image);
+    co_return;
+}
+
 sdbusplus::message::object_path Manager::startUpdate(
     sdbusplus::message::unix_fd image,
     ApplyTimeIntf::RequestedApplyTimes applyTime)
@@ -229,7 +737,7 @@ sdbusplus::message::object_path Manager::startUpdate(
     itemUpdater.createActivationWithApplyTime(id, objPath, applyTime);
 
     int newFd = dup(image);
-    ctx.spawn(processImage(newFd, applyTime, id, objPath));
+    ctx.spawn(processPFRImage(newFd, applyTime, id, objPath));
 
     return sdbusplus::message::object_path(objPath);
 }
diff --git a/update_manager.hpp b/update_manager.hpp
index caba541..c76364c 100644
--- a/update_manager.hpp
+++ b/update_manager.hpp
@@ -5,10 +5,40 @@
 #include <sdbusplus/async.hpp>
 #include <xyz/openbmc_project/Software/Update/server.hpp>
 
+#include <filesystem>
 #include <random>
 #include <string>
 #include <tuple>
 
+enum pfrImgPCType
+{
+    pfrCPLDUpdateCap = 0x00,
+    pfrPCHPFM = 0x01,
+    pfrPCHUpdateCap = 0x02,
+    pfrBMCPFM = 0x03,
+    pfrBMCUpdateCap = 0x04,
+    seamlessPCHUpdCap = 0x05,
+    pfrAFMUpdateCap = 0x06,
+    pfrCmpstCPLDUpdCap = 0x07,
+    pfrAddToUpdateAFM = 0x0a,
+    pfrPerDeviceAFM = 0x08,
+    pfrCmpstRetimerUpdCap = 0x09,
+    bmcSecureBootRotCap = 0xf1,
+    bmcSecureBootOTPCap = 0xf2
+};
+
+/* PFR image block 0 - As defined in HAS */
+struct pfrImgBlock0
+{
+    uint32_t tag;
+    uint8_t pcLength[4];
+    uint8_t pcType[4];
+    uint8_t reserved1[4];
+    uint8_t hash256[32];
+    uint8_t hash384[48];
+    uint8_t reserved2[32];
+} __attribute__((packed));
+
 namespace phosphor::software::updater
 {
 class ItemUpdater;
@@ -24,6 +54,9 @@ using ItemUpdaterIntf = phosphor::software::updater::ItemUpdater;
 using ApplyTimeIntf =
     sdbusplus::common::xyz::openbmc_project::software::ApplyTime;
 
+using AssociationList =
+    std::vector<std::tuple<std::string, std::string, std::string>>;
+
 /** @class Manager
  *  @brief Processes the image file from update D-Bus interface.
  *  @details The update manager class handles software updates and manages
@@ -58,6 +91,10 @@ class Manager : public UpdateIntf
                       std::string id, std::string objPath)
         -> sdbusplus::async::task<>;
 
+    auto processPFRImage(sdbusplus::message::unix_fd image,
+                         ApplyTimeIntf::RequestedApplyTimes applyTime,
+                         std::string id, std::string objPath)
+        -> sdbusplus::async::task<>;
     /* @brief The handler for the image processing failure  */
     void processImageFailed(sdbusplus::message::unix_fd image, std::string& id);
 
@@ -65,6 +102,13 @@ class Manager : public UpdateIntf
     std::mt19937 randomGen{static_cast<unsigned>(
         std::chrono::system_clock::now().time_since_epoch().count())};
 
+    int verifyImage(const std::filesystem::path imgPath, std::string& version,
+                    std::string& purposeString);
+
+    int verifyBmcFullFlashImage(const std::filesystem::path imgPath,
+                                std::string& version,
+                                std::string& purposeString);
+
     /** @brief D-Bus context */
     sdbusplus::async::context& ctx;
     /** @brief item_updater reference */
-- 
2.43.0

